examples_apps/run_state_learner.py

```diff
@@ -0,0 +1,432 @@
+# quantum circuit, and :math:`\ket{\psi_0}=\ket{0}` the initial state.
+#
+# Let's define the target state as the single photon state
+# :math:`\ket{\psi_t}=\ket{1}`:
+
+import numpy as np
+
+target_state = np.zeros([cutoff])
+target_state[1] = 1
+print(target_state)
+
+
+######################################################################
+# Using this target state, we calculate the fidelity with the state
+# exiting the variational circuit. We must use TensorFlow functions to
+# manipulate this data, as well as a ``GradientTape`` to keep track of the
+# corresponding gradients!
+#
+# We choose the following cost function:
+#
+# .. math:: C(\vec{\theta}) = \left| \langle \psi_t \mid U(\vec{\theta})\mid 0\rangle - 1\right|
+#
+# By minimizing this cost function, the variational quantum circuit will
+# prepare a state with high fidelity to the target state.
+
+
+def cost(weights):
+    # Mapping from the Strawberry Fields free parameters to
+    # the TensorFlow variables.
+    mapping = {
+        p.name: w
+        for (sf_layer, weights_layer) in zip(sf_params, weights * 1.0)
+        for (p, w) in zip(sf_layer, weights_layer)
+    }
+
+    # Run engine
+    state = eng.run(prog, args=mapping).state
+
+    # Extract the statevector
+    ket = state.ket()
+
+    # Compute the fidelity between the output statevector
+    # and the target state.
+    fidelity = tf.abs(tf.reduce_sum(tf.math.conj(ket) * target_state)) ** 2
+
+    # Objective function to minimize
+    cost = tf.abs(tf.reduce_sum(tf.math.conj(ket) * target_state) - 1)
+    return cost, fidelity, ket
+
+
+######################################################################
+# Now that the cost function is defined, we can define and run the
+# optimization as well as the quantum engine. Below, we choose the Adam
+# optimizer that is built into TensorFlow, as well as instantiating
+# the Strawberry Fields TensorFlow backend.
+
+opt = tf.keras.optimizers.Adam(learning_rate=lr)
+eng = sf.Engine("tf", backend_options={"cutoff_dim": cutoff})
+
+######################################################################
+# We then loop over all repetitions, storing the best predicted fidelity
+# value.
+
+fid_progress = []
+best_fid = 0
+
+
+for i in range(reps):
+    # reset the engine if it has already been executed
+    if eng.run_progs:
+        eng.reset()
+
+    with tf.GradientTape() as tape:
+        loss, fid, ket = cost(weights)
+
+    # Stores fidelity at each step
+    fid_progress.append(fid.numpy())
+
+    if fid > best_fid:
+        # store the new best fidelity and best state
+        best_fid = fid.numpy()
+        learnt_state = ket.numpy()
+
+    # one repitition of the optimization
+    gradients = tape.gradient(loss, weights)
+    opt.apply_gradients(zip([gradients], [weights]))
+
+    # Prints progress at every rep
+    if i % 1 == 0:
+        print("Rep: {} Cost: {:.4f} Fidelity: {:.4f}".format(i, loss, fid))
+
+
```

Comment:
![image](https://user-images.githubusercontent.com/49409390/80412028-317ff800-889b-11ea-8971-9d2b88fd0ad5.png)

On my local version it didn't seem to train :thinking:

Reply:
Which has a knock on effect to the rest of the results section

Reply:
```python
Rep: 150 Cost: 0.0955 Fidelity: 0.9607
Rep: 151 Cost: 0.0525 Fidelity: 0.9638
Rep: 152 Cost: 0.0595 Fidelity: 0.9653
Rep: 153 Cost: 0.0867 Fidelity: 0.9638
Rep: 154 Cost: 0.0447 Fidelity: 0.9646
Rep: 155 Cost: 0.0651 Fidelity: 0.9621
Rep: 156 Cost: 0.0905 Fidelity: 0.9603
Rep: 157 Cost: 0.0477 Fidelity: 0.9636
Rep: 158 Cost: 0.0631 Fidelity: 0.9662
Rep: 159 Cost: 0.0893 Fidelity: 0.9657
Rep: 160 Cost: 0.0459 Fidelity: 0.9670
Rep: 161 Cost: 0.0652 Fidelity: 0.9641
Rep: 162 Cost: 0.0920 Fidelity: 0.9615
Rep: 163 Cost: 0.0501 Fidelity: 0.9642
Rep: 164 Cost: 0.0596 Fidelity: 0.9657
Rep: 165 Cost: 0.0859 Fidelity: 0.9643
Rep: 166 Cost: 0.0442 Fidelity: 0.9649
Rep: 167 Cost: 0.0641 Fidelity: 0.9626
Rep: 168 Cost: 0.0893 Fidelity: 0.9610
Rep: 169 Cost: 0.0471 Fidelity: 0.9641
Rep: 170 Cost: 0.0620 Fidelity: 0.9662
Rep: 171 Cost: 0.0880 Fidelity: 0.9655
Rep: 172 Cost: 0.0456 Fidelity: 0.9668
Rep: 173 Cost: 0.0633 Fidelity: 0.9648
Rep: 174 Cost: 0.0894 Fidelity: 0.9627
Rep: 175 Cost: 0.0478 Fidelity: 0.9650
Rep: 176 Cost: 0.0604 Fidelity: 0.9659
Rep: 177 Cost: 0.0863 Fidelity: 0.9645
Rep: 178 Cost: 0.0452 Fidelity: 0.9653
Rep: 179 Cost: 0.0619 Fidelity: 0.9636
Rep: 180 Cost: 0.0870 Fidelity: 0.9620
Rep: 181 Cost: 0.0458 Fidelity: 0.9647
Rep: 182 Cost: 0.0613 Fidelity: 0.9662
Rep: 183 Cost: 0.0867 Fidelity: 0.9656
Rep: 184 Cost: 0.0449 Fidelity: 0.9670
Rep: 185 Cost: 0.0624 Fidelity: 0.9657
Rep: 186 Cost: 0.0880 Fidelity: 0.9640
Rep: 187 Cost: 0.0465 Fidelity: 0.9659
Rep: 188 Cost: 0.0603 Fidelity: 0.9660
Rep: 189 Cost: 0.0861 Fidelity: 0.9644
Rep: 190 Cost: 0.0460 Fidelity: 0.9651
Rep: 191 Cost: 0.0592 Fidelity: 0.9642
Rep: 192 Cost: 0.0839 Fidelity: 0.9631
Rep: 193 Cost: 0.0433 Fidelity: 0.9653
Rep: 194 Cost: 0.0619 Fidelity: 0.9666
Rep: 195 Cost: 0.0864 Fidelity: 0.9662
Rep: 196 Cost: 0.0447 Fidelity: 0.9679
Rep: 197 Cost: 0.0616 Fidelity: 0.9673
Rep: 198 Cost: 0.0873 Fidelity: 0.9655
Rep: 199 Cost: 0.0463 Fidelity: 0.9667
```

Reply:
Weird @trbromley, it trains fine for me locally (and matches @antalszava's output!) I wonder why it's not working for you 🤔 What version of TensorFlow do you have installed?

Reply:
:heavy_check_mark: Using TensorFlow==2.0.0:
```python
Rep: 195 Cost: 0.0212 Fidelity: 0.9909
Rep: 196 Cost: 0.0236 Fidelity: 0.9910
Rep: 197 Cost: 0.0212 Fidelity: 0.9910
Rep: 198 Cost: 0.0235 Fidelity: 0.9911
Rep: 199 Cost: 0.0211 Fidelity: 0.9910
```
:x: Using TensorFlow==2.1.0:
```python
Rep: 195 Cost: 1.0138 Fidelity: 0.0002
Rep: 196 Cost: 1.0133 Fidelity: 0.0002
Rep: 197 Cost: 1.0128 Fidelity: 0.0002
Rep: 198 Cost: 1.0126 Fidelity: 0.0002
Rep: 199 Cost: 1.0124 Fidelity: 0.0002
```
Interesting how the one that works doesn't match yours and @antalszava's results.

Reply:
That's super strange...

Reply:
Maybe it's something specific to my setup, do you get a problem with 2.1.0? And if so, isn't this something to worry about :sweat_smile:?

Reply:
@trbromley, can confirm that it doesn't train for me either when I upgrade to 2.1.0, but as soon as I downgrade to 2.0 it works again 🤔 I have no idea why this is the case

Reply:
Same for me as well with `2.1`.